File: sdk/typescript/rooch-sdk/src/utils/error.test.ts
// Copyright (c) RoochNetwork
// SPDX-License-Identifier: Apache-2.0

import { describe, it, expect } from 'vitest'
import {
  parseRoochErrorCode,
  parseRoochErrorSubStatus,
  ErrorCategory,
  getErrorCategoryName,
} from './error'

describe('err', () => {
  describe('parseRoochErrorCode', () => {
    it('should return the correct sub status', () => {
      const errorMessage = 'status ABORTED of type Execution with sub status 66537'
      expect(parseRoochErrorCode(errorMessage)).toBe(66537)
    })

    it('should return null if no sub status is found', () => {
      const errorMessage = 'status ABORTED of type Execution with no sub status'
      expect(parseRoochErrorCode(errorMessage)).toBeNull()
    })

    it('should return null if input is not a string', () => {
      const errorMessage = null
      expect(parseRoochErrorCode(errorMessage)).toBeNull()
    })
  })

  describe('parseRoochErrorSubStatus', () => {
    it('should return the correct sub status', () => {
      const errorMessage = 'status ABORTED of type Execution with sub status 66537'
      const subStatus = parseRoochErrorSubStatus(errorMessage)
      expect(subStatus).toBeDefined()
      expect(subStatus?.category).toBe(ErrorCategory.INVALID_ARGUMENT)
      expect(subStatus?.reason).toBe(1001)
    })

    it('should return null if no sub status is found', () => {
      const errorMessage = 'status ABORTED of type Execution with no sub status'
      expect(parseRoochErrorSubStatus(errorMessage)).toBeNull()
    })

    it('should return null if input is not a string', () => {
      const errorMessage = null
      expect(parseRoochErrorSubStatus(errorMessage)).toBeNull()
    })
  })

  describe('getErrorCategoryName', () => {
    it('should return the correct string representation of the enum', () => {
      expect(getErrorCategoryName(ErrorCategory.INVALID_ARGUMENT)).toBe('INVALID_ARGUMENT')
      expect(getErrorCategoryName(ErrorCategory.OUT_OF_RANGE)).toBe('OUT_OF_RANGE')
      expect(getErrorCategoryName(ErrorCategory.INVALID_STATE)).toBe('INVALID_STATE')
      expect(getErrorCategoryName(ErrorCategory.UNAUTHENTICATED)).toBe('UNAUTHENTICATED')
      expect(getErrorCategoryName(ErrorCategory.PERMISSION_DENIED)).toBe('PERMISSION_DENIED')
      expect(getErrorCategoryName(ErrorCategory.NOT_FOUND)).toBe('NOT_FOUND')
      expect(getErrorCategoryName(ErrorCategory.ABORTED)).toBe('ABORTED')
      expect(getErrorCategoryName(ErrorCategory.ALREADY_EXISTS)).toBe('ALREADY_EXISTS')
      expect(getErrorCategoryName(ErrorCategory.RESOURCE_EXHAUSTED)).toBe('RESOURCE_EXHAUSTED')
      expect(getErrorCategoryName(ErrorCategory.CANCELLED)).toBe('CANCELLED')
      expect(getErrorCategoryName(ErrorCategory.INTERNAL)).toBe('INTERNAL')
      expect(getErrorCategoryName(ErrorCategory.NOT_IMPLEMENTED)).toBe('NOT_IMPLEMENTED')
      expect(getErrorCategoryName(ErrorCategory.UNAVAILABLE)).toBe('UNAVAILABLE')
    })
  })
})

File: sdk/golang/rooch-sdk/utils/error_test.go
package utils

import (
	"testing"
)

func TestParseRoochErrorCode(t *testing.T) {
	t.Run("should return the correct sub status", func(t *testing.T) {
		errorMessage := "status ABORTED of type Execution with sub status 66537"
		result := ParseRoochErrorCode(errorMessage)
		if result == nil || *result != 66537 {
			t.Errorf("Expected 66537, got %v", result)
		}
	})

	t.Run("should return nil if no sub status is found", func(t *testing.T) {
		errorMessage := "status ABORTED of type Execution with no sub status"
		result := ParseRoochErrorCode(errorMessage)
		if result != nil {
			t.Errorf("Expected nil, got %v", result)
		}
	})

	t.Run("should return nil if input is nil", func(t *testing.T) {
		var errorMessage string
		result := ParseRoochErrorCode(errorMessage)
		if result != nil {
			t.Errorf("Expected nil, got %v", result)
		}
	})
}

func TestParseRoochErrorSubStatus(t *testing.T) {
	t.Run("should return the correct sub status", func(t *testing.T) {
		errorMessage := "status ABORTED of type Execution with sub status 66537"
		subStatus := ParseRoochErrorSubStatus(errorMessage)
		if subStatus == nil {
			t.Fatal("Expected non-nil subStatus")
		}
		if subStatus.Category != ErrorCategoryInvalidArgument {
			t.Errorf("Expected category INVALID_ARGUMENT, got %v", subStatus.Category)
		}
		if subStatus.Reason != 1001 {
			t.Errorf("Expected reason 1001, got %v", subStatus.Reason)
		}
	})

	t.Run("should return nil if no sub status is found", func(t *testing.T) {
		errorMessage := "status ABORTED of type Execution with no sub status"
		result := ParseRoochErrorSubStatus(errorMessage)
		if result != nil {
			t.Errorf("Expected nil, got %v", result)
		}
	})

	t.Run("should return nil if input is nil", func(t *testing.T) {
		var errorMessage string
		result := ParseRoochErrorSubStatus(errorMessage)
		if result != nil {
			t.Errorf("Expected nil, got %v", result)
		}
	})
}

func TestGetErrorCategoryName(t *testing.T) {
	t.Run("should return the correct string representation of the enum", func(t *testing.T) {
		testCases := []struct {
			category ErrorCategory
			expected string
		}{
			{ErrorCategoryInvalidArgument, "INVALID_ARGUMENT"},
			{ErrorCategoryOutOfRange, "OUT_OF_RANGE"},
			{ErrorCategoryInvalidState, "INVALID_STATE"},
			{ErrorCategoryUnauthenticated, "UNAUTHENTICATED"},
			{ErrorCategoryPermissionDenied, "PERMISSION_DENIED"},
			{ErrorCategoryNotFound, "NOT_FOUND"},
			{ErrorCategoryAborted, "ABORTED"},
			{ErrorCategoryAlreadyExists, "ALREADY_EXISTS"},
			{ErrorCategoryResourceExhausted, "RESOURCE_EXHAUSTED"},
			{ErrorCategoryCancelled, "CANCELLED"},
			{ErrorCategoryInternal, "INTERNAL"},
			{ErrorCategoryNotImplemented, "NOT_IMPLEMENTED"},
			{ErrorCategoryUnavailable, "UNAVAILABLE"},
		}

		for _, tc := range testCases {
			result := GetErrorCategoryName(tc.category)
			if result != tc.expected {
				t.Errorf("For category %v, expected %s, got %s", tc.category, tc.expected, result)
			}
		}
	})
} 